model/task_lifecycle.go

```diff
@@ -476,18 +476,20 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
 		grip.Warningf("Tried to mark task %s as finished twice", t.Id)
 		return nil
 	}
-	if detailsCopy.Status == evergreen.TaskSucceeded && t.MustHaveResults && !t.HasCedarResults {
-		// Results not in cedar, check the db.
+	if !t.HasCedarResults { // Results not in cedar, check the db.
 		count, err := testresult.Count(testresult.FilterByTaskIDAndExecution(t.Id, t.Execution))
 		if err != nil {
 			return errors.Wrap(err, "unable to count test results")
 		}
-		if count == 0 {
+		t.HasLegacyResults = utility.ToBoolPtr(count > 0) // cache if we even need to look this up in the future
+
+		if detailsCopy.Status == evergreen.TaskSucceeded && count == 0 && t.MustHaveResults {
```

Comment:
Can you please explain this if statement more? Why are we setting  detailsCopy.Status = evergreen.TaskFailed if detailsCopy.Status == evergreen.TaskSucceeded?

Reply:
This is existing logic that I just rearranged but the key part of this is `t.MustHaveResults`. If the yaml defines that the task must have test results, and we say that it's succeeded but there are no test results available, then this is a special failure, which we explain in the detailsCopy.Description